proposals/0370-modifiers.rst

```diff
@@ -167,8 +268,7 @@ Further examples:
 
 * Terms: Same as the example above.
 
-* Lambda expressions: ``\ x %Many -> ...`` or ``\ x %One -> ...``. This would be parsed but rejected, because
-  the new syntax applies only for lambda that bind a single, top-level variable: ``\ x y %One -> ...``.
+* Lambda expressions: ``\ %Many x -> ...``, ``\ %One x %Many y -> ...``.
```

Comment:
Updated to match the syntax (not yet implemented) from the linear types proposal.

I'm not sure if this is accepted, or if it would need to be `\ (%One x) (%Many y) ->`.

Reply:
I think it's fine (better, even) to require parentheses. It's more consistent with type annotation. Clearer to parse. It's the right way. Feel free to modify the linear types proposal accordingly.

Reply:
Hm, so on the question of what the specification *should* be: I don't feel strongly about parens being required when there's more than one argument. But when there's just one, I do feel like I prefer

    \ %One x ->

being allowed, rather than requiring

    \ (%One x) ->

and if the first one is allowed, then I'm not sure if it's feasible to require parens when there are multiple arguments. (That's actual not sure, not intended as "I don't think this is feasible".)

Reply:
The thing is, we already have:

```
> :t \ x::Int -> Int

<interactive>:1:4: error: [GHC-58481] parse error on input ‘::’
```

So it's worth being consistent. And it avoids a special case. Frankly this is a reasonable compromise. Even if `\ %1 x ->` does look better, I agree.

Reply:
Yeah, that makes sense. And if we do want to be more permissive later we can do that. (I definitely don't want a special case so that `\ %1 x ->` is allowed but `\ %1 x %1 y ->` isn't, but I don't have strong feelings about allowing `\ %1 x %1 y ->`.)

Ah, but, do you know how I'd amend the BNF in the linear types proposal? Currently the relevant section is:

> Lambda-bound variables can also be annotated with a multiplicity:
>
> ```haskell
> \ (%'One x) :: A -> x
> ```
>
> is the identity function at type `A ⊸ A`. A binder can be annotated with a multiplicity without a type like this
>
> ```haskell
> \ %'One x -> x
> ```
>
> This modifies the syntax entry for lambda expressions as follows
>
>     lpat -> [PREFIX_PERCENT btype] lpat
>
> where the btype after the % must be of kind Multiplicity (see below). The lpat must be a bare variable, or an error occurs.

And I don't know what the current BNF is, to know what it should be changed to.

Also, I think the first example there was probably a mistake? Since as you note, `\ x :: A ->` is currently forbidden, so it would be weird if `\ (%1 x) :: A ->` was allowed. So I guess I'll change these examples to

```haskell
\ (%1 x :: A) -> x
\ (%1 x) -> x
```

(I assume we don't want to require `\ ((%1 x) :: A) ->`.)

Reply:
Yeah, I think `(%1 x :: A)` is fine. It really doesn't look problematic from a parsing point of view. If it is, we can revisit.

For the BNF, imagine that the linear types proposal don't exist. So the modifiers proposal is retroactively a dependence of the linear types implementation.

Reply:
So, the problem I have is that I don't know what the BNF for the currently accepted syntax is. I found https://www.haskell.org/onlinereport/haskell2010/haskellch10.html but that doesn't allow `::` in patterns, that's a GHC extension. The extension is documented at https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/scoped_type_variables.html#pattern-type-sigs but not with BNF. (And it's not clear exactly which patterns allow signatures without being parenthesized and which don't; 

I could try to reverse engineer the BNF from the description, or from the parser code, but if there's an existing BNF written down somewhere, I'd be unlikely to match it exactly. Or I suppose I could try to look up where ScopedTypeVariables was first proposed and see if that has BNF?

But... I kinda feel like most people writing proposals probably don't do that kind of thing, and either know somewhere to look up the currently supported BNF, or just kinda guess?

If I'm guessing, then... well, somewhat embarrassingly, I now see that the current modifiers proposal has `lpat ::= modifiers lpat | ...`, and that feels like it might already be exactly what we want here. And I think it's the same BNF change as in the linear types proposal, just with different BNF syntax. Now that I know roughly what `lpat` is (it turns out it's not "the pattern inside a lambda", which is an `apat`), I see that yeah, the BNF given in the current LT proposal doesn't support the syntax examples in it.

All of which is to say, I currently think the changes I need to make are

* Update the examples in the linear types and modifiers proposals, showing that the syntax inside lambdas is `\ (%1 a) ->` and `\ (%1 a :: A) ->`.
* Remove the bit about "The `lpat` must be a bare variable" in the LT proposal, since that's contradicted by e.g. support for `let %1 !(x, y) = ...`. (It's kind of confusing that the BNF for patterns-in-general is described in a section talking specifically about lamba variables, but I think I'm not going to try fixing that.)

Reply:
Done.

Reply:
> It's kind of confusing that the BNF for patterns-in-general is described in a section talking specifically about lamba variables, but I think I'm not going to try fixing that

No need, indeed.

> So, the problem I have is that I don't know what the BNF for the currently accepted syntax is.

What we do, in practice, is provide a diff with respect, either to the last Haskell report when it's still close enough to the current parser (but as the report is getting older, it becomes less and less relevant). Or simply open the parser proper and define a diff to that (in stylised BNF form).

Thanks for doing it.